src/lazy.rs

```diff
@@ -0,0 +1,29 @@
+use core::cell::UnsafeCell;
+
+pub struct LazyCell<T> {
+    contents: UnsafeCell<Option<T>>,
+}
+impl<T> LazyCell<T> {
+    pub fn new() -> LazyCell<T> {
+        LazyCell {
+            contents: UnsafeCell::new(None),
+        }
+    }
+
+    pub fn borrow_with(&self, closure: impl FnOnce() -> T) -> &T {
+        unsafe {
+            // First check if we're already initialized...
+            let ptr = self.contents.get();
+            if let Some(val) = &*ptr {
+                return val;
+            }
+            // Note that while we're executing `closure` our `borrow_with` may
+            // be called recursively. This means we need to check again after
+            // the closure has executed. For that we use the `get_or_insert`
+            // method which will only perform mutation if we aren't already
+            // `Some`.
+            let val = closure();
+            (*ptr).get_or_insert(val)
```

Comment:
```suggestion
            (*self.contents.get()).get_or_insert(val)
```

If `borrow_with` is called from inside the closure, then `ptr` is invalidated according to stacked borrows I think.

Reply:
I don't believe this is necessary since the value of `ptr` doesn't change, and borrows into it are temporary. If miri or something throws an error here it can of course be changed.

Reply:
`ptr` is of course a raw pointer.

Reply:
@bjorn3 I'm taking your comment to mean that this conversation is resolved.